### PR TITLE
feat(gitlab-application): Add resource for gitlab application, data source for current user

### LINKED ISSUE #321: Add support for Github Applications
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

It would be nice to be able to have support for [GitLab Applications](https://docs.gitlab.com/ee/api/applications.html).

### New or Affected Resource(s)

* gitlab_application

### Potential Terraform Configuration

```hcl
resource "gitlab_application" "example" {
  name = "My App"
  callback_uri = "https://example.com/login/callback"
  scopes = [
    "api"
  ]
  confidential = true
}
```

Not shown here but the `secret` from the exported attributes should probably be encrypted using a PGP key like for https://www.terraform.io/docs/providers/aws/r/iam_access_key.html.

### References

- GitLab Actions: https://docs.gitlab.com/ee/api/applications.html
- Go GitLab client API: https://godoc.org/github.com/xanzy/go-gitlab#Application

### PR DESCRIPTION
**Pull request by vymarkov**
Fixes #321 

This PR adds the ability to create [an application](https://docs.gitlab.com/ee/api/applications.html), data source for [gitlab current user](https://docs.gitlab.com/ce/api/users.html#list-current-user-for-normal-users) to don't fail in case if gitlab isn't self-hosted, see below the terraform configuration for this case.

```hcl
data "gitlab_current_user" "me" {
}

// https://github.com/terraform-providers/terraform-provider-gitlab/issues/321
resource "gitlab_application" "grafana" {
  // create the application if user has admin role, 
  // otherwise Gitlab API rejects requests with 403 http status code
  count = data.gitlab_current_user.me.is_admin ? 1 : 0

  name = "grafana"
  redirect_uri = [
    "http://localhost:8080/auth/gitlab/callback",
    "http://localhost:8080/auth/gitlab/callback2"
  ]
  scopes = ["openid", "email", "profile"]
}

output "me" {
  value = data.gitlab_current_user.me
}

output "grafana_app_secret" {
  value = {
    app_id     = gitlab_application.grafana[0].application_id
    app_secret = gitlab_application.grafana[0].secret
  }
  sensitive = true
}

```

Gitlab API has limitations that don't allow to update the existing application, thus any property changes will force deleting and creating new one.

### CODE REVIEW

**Comment by roidelapluie** [comment]
Can you remove the changes of the Makefile from this PR? Thanks

**Comment by vymarkov** [comment]
> Can you remove the changes of the Makefile from this PR? Thanks

done :)

**Comment by armsnyder** [comment]
Can you purge the terraform-provider-gitlab binary file from this PR and its commits?

File: gitlab/resource_gitlab_application.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		appsLen = len(apps)
+		page = page + 1
+	}
+
+	log.Printf("[DEBUG] failed to read gitlab application %s", appID)
+	d.SetId("")
+	return nil
+}
+
+func resourceGitlabApplicationUpdate(d *schema.ResourceData, meta interface{}) error {
```
Comment (armsnyder) [inline gitlab/resource_gitlab_application.go:127]: If applications cannot be updated, can you remove the `Update` function altogether and make all of the attributes `ForceNew: true` in the schema?

**Comment by vymarkov** [comment]
I'm going to pick up this task next week

**Comment by timofurrer** [comment]
Duplicate of #417 . I'm going to close this one and let's follow up with it on #417.

**Comment by armsnyder** [comment]
It sounds like the main attraction of this PR was to add a `gitlab_current_user` datasource, which is not in #417.

**Pull request closed without merging**